code/game/objects/structures/training_machine.dm

```diff
@@ -0,0 +1,310 @@
+
+/**
+  * Machine that runs around wildly so people can practice clickin on things
+  *
+  * Can have a mob buckled on or a obj/item/target attached. Uses timers to move, which is necessary because it's movespeed
+  * needs to both be variable and potentially faster than SSFastProcess's tick speed, so hey this is what we got. 
+  * As a note, there generally shouldn't be too many timers in existance at a time, so there will only ever be
+  * one or two of these on the station.
+  */
+/obj/structure/training_machine
+	name = "AURUMILL-Brand MkII. Personnel Training Machine"
+	desc = "Used for combat training simulations. Accepts standard training targets. A pair of buckling straps are attached."
+	icon = 'icons/obj/objects.dmi'
+	icon_state = "training_machine"
+	can_buckle = TRUE
+	buckle_lying = 0
+	max_integrity = 200
+	///Is the machine moving? Setting this to FALSE will automatically call stop_moving()
+	var/moving = FALSE
+	///The distance the machine is allowed to roam from its starting point
+	var/range = 1
+	///A random spot within range that the trainer is trying to reach
+	var/turf/target_position
+	///The turf the machine was on when it was activated
+	var/turf/starting_turf
+	///How fast the machine moves. Cannot be higher than MAX_SPEED
+	var/move_speed = 1
+	///Reference to a potentially attached object, either a target, trainer toolbox, or syndicate toolbox
+	var/obj/item/attached_item
+	///Time between attacks when emagged
+	var/attack_cooldown = 50
+	///Helper for timing attacks when emagged
+	var/last_attack_time = 0
+
+/obj/structure/training_machine/Destroy()
+	remove_attached_item(throwing = TRUE)
+	explosion(src, 0,0,1, flame_range = 2)
+	return ..()
+
+/obj/structure/training_machine/ui_state(mob/user)
+	return GLOB.physical_state
+
+/obj/structure/training_machine/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "TrainingMachine", name)
+		ui.open()
+
+/obj/structure/training_machine/ui_data(mob/user)
+	var/list/data = list()
+	data["range"] = range
+	data["movespeed"] = move_speed
+	data["moving"] = moving
+	return data
+
+/obj/structure/training_machine/ui_act(action, params)
+	if(..())
+		return
+	if (moving && obj_flags & EMAGGED)
+		visible_message("<span class='warning'>The [src]'s control panel fizzles slightly.</span>")
+		return
+	switch(action)
+		if("toggle")
+			toggle()
+			. = TRUE
+		if("range")
+			var/range_input = params["range"]
+			range = clamp(range_input, MIN_RANGE, MAX_RANGE)
+			. = TRUE
+		if("movespeed")
+			var/range_input = params["movespeed"]
+			move_speed = clamp(range_input, MIN_SPEED, MAX_SPEED)
+			. = TRUE
+
+/obj/structure/training_machine/attack_hand(mob/user)
+	ui_interact(user)
+
+/obj/structure/training_machine/attackby(obj/item/target, mob/user)
+	. = ..()
+	if (user.a_intent == INTENT_HARM)
+		return
+	if (!istype(target, /obj/item/training_toolbox) && !istype(target, /obj/item/target))
+		return
```

Comment:
You might want to either handle parent returning a value not to continue the attack or to only call parent when your desired interaction options are unavailable.